design-proposals/volume-update-strategy.md

```diff
@@ -0,0 +1,234 @@
+
+# Design
+
+This proposal introduces a new field in the VM spec to represent the volume update strategy called `updateVolumesStrategy`. This value is a unique field per VM, it can takes the values:
+ * `replacement` when the update of the volume aims to substitute the volume with another upon a VM restart
+ * `migration` when the update of the volume wants to migrate a volume to another one performing a full-copy of the content
+
+If no value is specified, the default behavior is `replacement`.
+
+Possibly, in the future, we could add a strategy for hotplugged volumes when the update wants to immediately add a new volume to a running VM.
+
+## API examples
+
+#### VM Spec API update
+
+Original VM spec
+```yaml
+  spec
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: src-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: src-pvc2
+  - name: vol3
+     persistentVolumeClaim:
+    claimName: src-pvc3
+```
+
+Update VM spec:
+```yaml
+spec:
+  updateVolumesStrategy: migration
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: dst-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: dst-pvc2
+  - name: vol3
+    persistentVolumeClaim:
+      claimName: src-pvc3
+```
+
+In this, example the user is able to update and migrate the first and second volume while the third one remains untouched.
+
+## Flow
+
+The volume migration update strategy follow the same [update sequence](https://github.com/kubevirt/community/blob/main/design-proposals/cpu-hotplug.md#phase-2-triggering-a-live-migration-when-the-number-of-sockets-changes-to-adjust-the-virt-launcher-cpu-resources) proposed for memory and cpu hotplug.
+
+Sequence of events if the `updateVolumesStrategy` is migration:
+
+* the user updates the volumes on a vm.spec.template.spec.domain.volumes
+* the VM controller determines the volumes which needs to be updated
+* the VM controller validates if the volumes can be migrated
+* the VM controller updates the volumes accordingly in the active vmi.Spec
+* the VMI controller writes the required volumes to migrate in the vmi.Status and set the update condition in the VMI status
+* the VMI workload-update controller detects the update condition `VolumesChange` and triggers the VMI migration. When it creates the VMI migration object it also set the label `kubevirt.io/volume-update-migration: vm-name`.
+* the Migration controller constructs a new target pod with the desired volumes
+* virt-launcher updates the volumes to migrate in the migration request for LibVirt
+* the VMI live migrates
+* the VMI status is updated to reflect that the volume update action have completed and the update condition is removed
+
+This feature depends on the `vmRolloutStrategy` which needs to be set to `LiveUpdate`. On the long term, this should become the default behavior.
+
+The label `kubevirt.io/volume-update` can be used to filter and identify the VM migrations triggered by a volume update.
```

Comment:
`kubevirt.io/volume-update`

I'm not sure baout the semantics, can the label be more clear, i.e.
`kubevirt.io/consequence-of-volume-update`
or
`kubevirt.io/volume-update-in-progress`
or
`kubevirt.io/due-to-volume-update`

Reply:
ok, I would go for `kubevirt.io/volume-update-in-progress`